core/internal/viewstate/view-state.go

```diff
@@ -0,0 +1,272 @@
+// normal mode. The returned view number is guaranteed to denote the
+// current active view until the returned release function is invoked.
+//
+// WaitAndHoldView waits for specific active view and defers view
+// change. This method resembles WaitAndHoldActiveView, but only waits
+// for a specific active view number supplied through the parameter
+// view. It immediately returns false once it is detected that the
+// supplied view number will never become active. Otherwise, it
+// returns true as soon as the awaited view number is active. In that
+// case, the view will stay active until the returned release function
+// is invoked.
+//
+// RequestViewChange handles synchronization of view change requests.
+// If the supplied view number is greater than the last requested and
+// the active view number then it returns true and records the view
+// number as last requested; otherwise, false is returned.
+//
+// StartViewChange synchronizes beginning of view change. If in the
+// normal mode and the supplied value is higher than the active view
+// number then it switches to the view-change mode with the expected
+// new view number equal to the supplied value. If already in the
+// view-change mode and the supplied value is higher than the expected
+// new view number then the expected new view number is updated.
+// Before switching to the view-change mode, this method blocks any
+// new invocation of WaitAndHoldActiveView or WaitAndHoldView and
+// waits for all release functions returned from WaitAndHoldActiveView
+// or WaitAndHoldView to be invoked. It returns true if the replica
+// switched to the view-change mode or the expected new view number
+// was increased; otherwise false. If true is returned, any new
+// invocation of StartViewChange or FinishViewChange is blocked until
+// the returned release function is invoked.
+//
+// FinishViewChange synchronizes completion of view change. If the
+// replica is in the normal mode then it first begins a view change,
+// same as StartViewChange. If the supplied value equals to the
+// expected new view number, it returns true together with a release
+// function; otherwise false. If true is returned, any new invocation
+// of StartViewChange or FinishViewChange is blocked until the
+// returned release function is invoked. Invocation of the release
+// function switches the replica back to the normal mode with the
+// active view number equal to the supplied value, and unblocks any
+// waiting WaitAndHoldActiveView or WaitAndHoldView invocation.
+type State interface {
+	WaitAndHoldActiveView() (view uint64, release func())
+	WaitAndHoldView(view uint64) (ok bool, release func())
+	RequestViewChange(newView uint64) bool
+	StartViewChange(newView uint64) (ok bool, release func())
+	FinishViewChange(newView uint64) (ok bool, release func())
+}
+
+// In the normal mode, if lastNewView == lastActiveView. Otherwise, in
+// the view-change mode, and lastNewView > lastActiveView.
```

Comment:
This comment is very helpful, and it would be more readable if you introduce some helper functions named (for example) `isNormalMode`/`isViewChangeMode`. Checking state conditions in raw numbers are sometimes hard to understand what the conditions mean.

Reply:
Thank you for your suggestion. Let me think of some helper function to make the code read easy. So far I can see that `isViewChangeMode` would not be used anywhere. But other kind of helper might be useful.

Reply:
What I can suggest are the following helpers.

```go
func (s *viewState) inNormalMode() bool {
	return s.lastActiveView == s.lastNewView
}

func (s *viewState) inView(view uint64) bool {
	return s.inNormalMode() && view == s.lastActiveView
}

func (s *viewState) isNewView(view uint64) bool {
	return view > s.lastActiveView
}

func (s *viewState) isNewViewCandidate(view uint64) bool {
	return view > s.lastNewView
}

func (s *viewState) isPossibleNewView(view uint64) bool {
	return s.isNewView(view) && view >= s.lastNewView
}
```

[Here](https://github.com/sergefdrv/minbft/compare/add-view-state..sergefdrv:add-view-state-2) is how the code would change. Does it read better?

Reply:
Very nice, thank you very much.